src/terminal/adapter/DispatchTypes.hpp

```diff
@@ -89,19 +89,117 @@ namespace Microsoft::Console::VirtualTerminal
         uint64_t _idAccumulator = 0;
         size_t _idShift = 0;
     };
+
+    class VTParameter
+    {
+    public:
+        constexpr VTParameter() noexcept :
+            _value{ -1 }
+        {
+        }
+
+        constexpr VTParameter(const size_t rhs) noexcept :
+            _value{ gsl::narrow_cast<decltype(_value)>(rhs) }
+        {
+        }
+
+        constexpr bool has_value() const noexcept
+        {
+            // A negative value indicates that the parameter was omitted.
+            return _value >= 0;
+        }
+
+        constexpr size_t value() const noexcept
+        {
+            return _value;
+        }
+
+        constexpr size_t value_or(size_t defaultValue) const noexcept
+        {
+            return has_value() ? _value : defaultValue;
+        }
+
+        template<typename T, std::enable_if_t<sizeof(T) == sizeof(size_t), int> = 0>
+        constexpr operator T() const noexcept
+        {
+            // For most selective parameters, omitted values will default to 0.
+            return static_cast<T>(value_or(0));
+        }
+
+        constexpr operator size_t() const noexcept
+        {
+            // For numeric parameters, both 0 and omitted values will default to 1.
+            return has_value() && _value != 0 ? _value : 1;
+        }
+
+    private:
+        std::make_signed<size_t>::type _value;
+    };
+
+    class VTParameters
+    {
+    public:
+        constexpr VTParameters() noexcept
+        {
+        }
+
+        constexpr VTParameters(const VTParameter* ptr, const size_t count) noexcept :
+            _values{ ptr, count }
+        {
+        }
+
+        constexpr VTParameter at(const size_t index) const noexcept
+        {
+            // If the index is out of range, we return a parameter with no value.
+            return index < _values.size() ? _values[index] : VTParameter{};
+        }
+
+        constexpr bool empty() const noexcept
+        {
+            return _values.empty();
+        }
+
+        constexpr size_t size() const noexcept
+        {
+            // We always return a size of at least 1, since an empty parameter
+            // list is the equivalent of a single "default" parameter.
+            return std::max<size_t>(_values.size(), 1);
+        }
+
+        VTParameters subspan(const size_t offset) const noexcept
+        {
+            const auto subValues = _values.subspan(offset);
+            return { subValues.data(), subValues.size() };
+        }
+
+        bool for_each(const std::function<bool(const VTParameter)> predicate) const

```

Comment:
```suggestion
		template <typename TLambda>
        bool for_each(TLambda&& predicate) const
```

This has slightly weaker guarantees at compile time than `std::function` about arguments (though they'll be enforced on 182 when you try to call it), but eliminates _all_ of the overhead of `std::function`'s type-erasing virtual dispatcher. There's a minor cost in that one template instance is emitted for every different lambda given that each lambda is a new type, but COMDAT folding and ICF should more than take care of it.

The vast majority of these cases may even be fully inlinable once the compiler sees through the trick. I don't believe that the same can be said about `std::function`.

Reply:
Yeah, that does look better. And you're right about it inlining. Thanks for the suggestion.